File: 2025/02 February/23 Next Greater Element/main.py
# Link: https://www.geeksforgeeks.org/problems/next-larger-element-1587115620/1


class Solution:
    def nextLargerElement(self, arr):
        n = len(arr)
        ans = [0] * n

        st = []
        for i in range(n - 1, -1, -1):
            while st and arr[i] >= st[-1]:
                st.pop()
            if not st:
                ans[i] = -1
            else:
                ans[i] = st[-1]
            st.append(arr[i])
        return ans


if __name__ == "__main__":
    object = Solution()
    arr = [1, 3, 2, 4]
    result = object.nextLargerElement(arr)
    print(" ".join(map(str, result)))

    arr = [6, 8, 0, 1, 3]
    result = object.nextLargerElement(arr)
    print(" ".join(map(str, result)))

    arr = [10, 20, 30, 50]
    result = object.nextLargerElement(arr)
    print(" ".join(map(str, result)))

    arr = [50, 40, 30, 10]
    result = object.nextLargerElement(arr)
    print(" ".join(map(str, result)))

File: 2025/02 February/23 Next Greater Element/main.cpp
// Link: https://www.geeksforgeeks.org/problems/next-larger-element-1587115620/1

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

class Solution
{
public:
    vector<int> nextLargerElement(vector<int> &arr)
    {
        int n = arr.size();
        vector<int> ans(n, 0);

        stack<int> st;
        for (int i = n - 1; i >= 0; i--)
        {
            while (!st.empty() && arr[i] >= st.top())
            {
                st.pop();
            }
            if (st.empty())
            {
                ans[i] = -1;
            }
            else
            {
                ans[i] = st.top();
            }
            st.push(arr[i]);
        }
        return ans;
    }
};

int main()
{
    Solution object;
    vector<int> arr = {1, 3, 2, 4};
    vector<int> result = object.nextLargerElement(arr);
    for (int i = 0; i < result.size(); i++)
    {
        cout << result[i] << " ";
    }
    cout << endl;

    arr = {6, 8, 0, 1, 3};
    result = object.nextLargerElement(arr);
    for (int i = 0; i < result.size(); i++)
    {
        cout << result[i] << " ";
    }
    cout << endl;

    arr = {10, 20, 30, 50};
    result = object.nextLargerElement(arr);
    for (int i = 0; i < result.size(); i++)
    {
        cout << result[i] << " ";
    }
    cout << endl;

    arr = {50, 40, 30, 10};
    result = object.nextLargerElement(arr);
    for (int i = 0; i < result.size(); i++)
    {
        cout << result[i] << " ";
    }
    cout << endl;
    return 0;
}